### PR TITLE
handle file move (rename) and deletion

### LINKED ISSUE #230: [editor] Renaming a file should rename a corresponding editor
Consider also what to do with an underlying file resource and monaco model.

### LINKED ISSUE #560: [editor] deleting a file should be reflected in the corresponding editor
If a file is opened in an editor, deleting it ATM leaves the corresponding editor opened. It should probably be closed.

See also #230 . 

This has the side-effect of breaking the layout restoration, if the Theia client is closed with that non-existing file opened in an editor : 

``` javascript
localhost/:1 Uncaught (in promise) A resource provider for 'file:///tmp/a/theia/examples/browser/aaa.txt' is not registered.
Promise rejected (async)
(anonymous) @ shell-layout-restorer.ts:78
ShellLayoutRestorer.inflate @ shell-layout-restorer.ts:73
(anonymous) @ shell-layout-restorer.ts:28
Promise resolved (async)
ShellLayoutRestorer.onStart @ shell-layout-restorer.ts:25
FrontendApplication.startContributions @ frontend-application.ts:110
FrontendApplication.start @ frontend-application.ts:67
start @ index.js:29
Promise resolved (async)
Object.defineProperty.value @ index.js:48
__webpack_require__ @ bootstrap a28e50b24baf5c30315c:688
Object.defineProperty.value @ bootstrap a28e50b24baf5c30315c:789
(anonymous) @ bootstrap a28e50b24baf5c30315c:789
```

### LINKED ISSUE #1191: Moving a file open in an editor doesn't work well
If you have a file open in an editor and move it in a different directory using the navigator, it will re-create an empty file at the original location and open this one in the editor.  Ideally, the editor would keep the moved file open.

### LINKED ISSUE #1307: [editor][question] Auto-save can revive a deleted file in the WS
I hit this when I was trying to reproduce a defect in the navigator; is it a bug or a feature?
gh-1307

### LINKED ISSUE #1476: [fs] Implement smart multi-file deletion
Once #20 will be merged to the master one can select multiple files in the navigator, and can delete multiple files at the same time.

Make the multi-file deletion smarter than just delete each file one by one.
For instance:
 - When a folder and any of its contained files are deleted => delete just the folder.
 - When deleting all the files from a directory => delete the directory instead.

The corresponding conversation is [here](https://github.com/theia-ide/theia/pull/1438#discussion_r173075444)

### LINKED ISSUE #1574: Error when deleting a file which is opened in the editor.
When deleting a file:
```
Uncaught (in promise) Error: Request getFileStat failed with message: Cannot find file under the given URI. URI: file:///Users/akos.kitta/Desktop/bar/Untitled.txt.
    at new ResponseError (messages.js:46)
    at handleResponse (main.js:430)
    at processMessageQueue (main.js:258)
    at main.js:242
    at run (setImmediate.js:40)
    at runIfPresent (setImmediate.js:69)
    at onGlobalMessage (setImmediate.js:109)
```

Steps to reproduce:
- Create a file.
- Delete it.

Since the stack trace is absolutely useless, [here is a hint](https://github.com/theia-ide/theia/blob/c145b1533c69fb2e560099b85bc5b12c18bd94b8/packages/filesystem/src/browser/file-resource.ts#L81) to the source of the problem.

### LINKED ISSUE #1673: Implement Path.relative(to)
We already use this internally for Git, when getting the [repository relative path](https://github.com/theia-ide/theia/blob/91cb3123cdaed1794b49b576a8e5ab1354c56082/packages/git/src/browser/git-repository-tracker.ts#L88) for a file.

```typescript
const path = uri.path.toString();
 if (!path.startsWith(repositoryPath)) {
  return undefined;
 }
const relativePath = path.substr(repositoryPath.length);
return relativePath[0] === '/' ? relativePath.substr(1) : relativePath;
```

Also see: [`path.relative(from, to)`](https://nodejs.org/api/path.html#path_path_relative_from_to). In our case `from` would be `this`.

Add tests for Windows and OS X/Linux.

### LINKED ISSUE #1735: [git/dirty-diff] Error on backend when managed file w/ untracked changes is deleted
another variation of a case reported here: https://github.com/theia-ide/theia/issues/1475#issuecomment-372952304

The event of [untracked files being deleted](https://github.com/theia-ide/theia/issues/1475#issuecomment-372975176) is not handled. (TODOs are updated.)

@kittaakos pointed out how to detect such an event, [here](https://github.com/theia-ide/theia/issues/1475#issuecomment-372981499).

### LINKED ISSUE #2432: [editor] Restoring the visible position (maybe the selection) would be awesome when restoring the editor
VS Code can do it; it even restores the selected ranges.

### LINKED ISSUE #2538: [navigator] right click on empty area does not context menu anymore
It is a regression before the empty have belonged to the project.

In single root mode, it should belong to the project as before. In multi-root it should allow to add and remove roots.

### PR DESCRIPTION
**Pull request by akosyakov**
fix #2538, fix #1574, fix #1191, fix #560, fix #230, fix #2432, fix #1673, fix #1735, fix #1476, fix #1307

### CODE REVIEW

File: packages/core/src/common/path.ts [hunk 1]
```diff
@@ -123,4 +123,20 @@ export class Path {
...
+        if (value === value2) {
+            return true;
+        }
+        if (!value || !value2) {
+            return false;
+        }
+        if (this.root && Path.isDrive(this.root.base)) {
+            return value2.startsWith(value);
+        }
+        return value2.toLowerCase().startsWith(value.toLowerCase());
```
Comment (marechal-p) [inline packages/core/src/common/path.ts]: What about case-sensitive filesystems? Isn't there a corner case like `/home/a/foo` and `/home/a/Foo`?

Comment (marechal-p) [reply packages/core/src/common/path.ts]: Oh ok so you test for case-sensitive systems first, and for non-case sensitive systems after?
Isn't `Path.isDrive(this.root.base)` meant to detect Windows though?

Comment (akosyakov) [reply packages/core/src/common/path.ts]: you are right, it makes sense only in linux, I've not considered other since it is hard to decide on the frontend, I will assume that our fs api already returns normalized URIs.

File: packages/core/src/common/path.ts [hunk 2]
```diff
@@ -123,4 +123,24 @@ export class Path {
...
+    }
+
+    isEqualOrParent(path: Path): boolean {
+        if (this.raw === path.raw) {
+            return true;
+        }
+        if (!this.raw || !path.raw) {
+            return false;
+        }
+        return path.raw.startsWith(this.raw);
```
Comment (marechal-p) [inline packages/core/src/common/path.ts]: Still wondering about the path... Wouldn't `/home/me/test/file.txt` and `/home/me/t` cause problems?
In this case it would assume that `/home/me/t` is the parent of `/home/me/test/file.txt`, right?

**Comment by akosyakov** [comment]
@sve It's ready for review.

Builds are failing because of unreachable JDT LS: https://travis-ci.org/theia-ide/theia/jobs/415052039#L759 cc @kittaakos

**Comment by akosyakov** [comment]
There is still an issue that nsfw only reports ADD, DELETE events for files on top-level directories. E.g creating/deleting a file in src folder of any package in Theia is never worked.  Strangely if a file is moved or its content changed then events are reported always.

File: packages/workspace/src/browser/workspace-delete-handler.ts [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+    @inject(WorkspaceService)
+    protected readonly workspaceService: WorkspaceService;
+
+    isVisible(uris: URI[]): boolean {
+        if (!uris.length) {
+            return false;
+        }
+        const rootUris = this.workspaceService.tryGetRoots().map(root => new URI(root.uri));
+        return !rootUris.some(rootUri => uris.some(uri => uri.isEqualOrParent(rootUri)));
```
Comment (svenefftinge) [inline packages/workspace/src/browser/workspace-delete-handler.ts]: Looks like the command would be visible if one of the uris has one of the roots as parent. 
Didn't you want to check all, i.e.
```ts
return uris.every(uri => rootUris.some(rootUri => uri.isEqualOrParent(rootUri)));
```

File: packages/core/src/browser/navigatable.ts [hunk 4]
```diff
@@ -15,18 +15,59 @@
...
  * (e.g. editor, image viewer, diff editor, etc.) has to implement this interface and provide the file uri on demand.
  * No additional registration is needed.
  */
 export interface Navigatable {
+    /**
+     * Return an underlying file uri.
+     */
     getTargetUri(): URI | undefined;
+    /**
+     * Return a source uri for the given file URI.
```
[the remark this answers is not in the document]
Comment (svenefftinge) [reply packages/core/src/browser/navigatable.ts]: After reading the PR it is still not clear what the purpose of this method is. It seems to be a way to compute a URI based on a file move event. But why can'T we just use the targetURI from the event? Why does it need to get the scheme from the widget?

File: packages/editor/src/browser/editor-widget.ts [hunk 5]
```diff
@@ -39,7 +39,26 @@ export class EditorWidget extends BaseWidget implements SaveableSource, Navigata
...
+        const { uri } = this.editor;
+        if (DiffUris.isDiffUri(uri)) {
+            const [left, right] = DiffUris.decode(uri);
+            if (left.scheme === 'file') {
+                return left;
+            }
+            if (right.scheme === 'file') {
+                return right;
+            }
+            return undefined;
```
[the remark this answers is not in the document]
Comment (svenefftinge) [reply packages/editor/src/browser/editor-widget.ts]: I noticed this is just moved from another place. So it used to be like that. So the question is basically unrelated to this PR.

File: packages/workspace/src/browser/workspace-delete-handler.ts (hunk 3 above)
Comment (svenefftinge) [reply packages/workspace/src/browser/workspace-delete-handler.ts]: Do you have any cases in mind, where a URI would not be a child of a workspace root? And if so why would you want to disable the command?

File: packages/markers/src/browser/marker-manager.ts [hunk 6]
```diff
@@ -115,14 +115,14 @@ export abstract class MarkerManager<D extends object> {
 
     @postConstruct()
     protected init(): void {
-        this.fileWatcher.onFilesChanged(changes => {
-            for (const change of changes) {
-                if (change.type === FileChangeType.DELETED) {
-                    const uriString = change.uri.toString();
+        this.fileWatcher.onFilesChanged(event => {
+            for (const uriString of this.uri2MarkerCollection.keys()) {
```
[the remark this answers is not in the document]
Comment (akosyakov) [reply packages/markers/src/browser/marker-manager.ts]: If a folder is deleted, events for its children are not reported, but markers for all files belonging to the folder should be cleaned. It is the common issue with our code before this PR. `FileChangeEvent.isDeleted` takes it into the consideration: https://github.com/theia-ide/theia/pull/2544/files#diff-711943eab1ae2c1062adf6c4a472a25eR40

File: packages/workspace/src/browser/workspace-delete-handler.ts (hunk 3 above)
Comment (akosyakov) [reply packages/workspace/src/browser/workspace-delete-handler.ts]: > Looks like the command would be visible if one of the uris has one of the roots as parent.

`uri.isEqualOrParent(rootUri)` - it checks that no root will be deleted, not that a file should belong to a root.

File: packages/core/src/browser/navigatable.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (akosyakov) [reply packages/core/src/browser/navigatable.ts]: An idea was to have an inverse operation to `getTargetUri`:
- `getTargetUri` gets an underlying file URI from the widget URI, e.g. from a diff URI for the editor widget.
- `getSourceUri` creates a new widget URI based on a given file URI and a current widget URI.

In the case of the diff editor it creates a diff URI. In the case of the preview, it preserves a query if it was opened via a link. A target URI from a move event does not have this information. 

I've added it because it would be correct to replace only a path of URI, but it also makes following (esoteric) cases possible:
source

I am not sure about naming. `getTargetUri` was named like that to avoid colliding with `getUri` which is different. `getSourceUri` just to have a symmetric name.

[the remark this answers is not in the document]
Comment (svenefftinge) [reply packages/core/src/browser/navigatable.ts]: Understood, in that case the name should be something like `createMoveToURI` and have the explanation above as its doc comment :)

File: packages/workspace/src/browser/workspace-delete-handler.ts (hunk 3 above)
Comment (svenefftinge) [reply packages/workspace/src/browser/workspace-delete-handler.ts]: Maybe it's worth have a short inline comment about the intend as it wasn't clear (to me at least :)).

**Comment by akosyakov** [comment]
@svenefftinge all your comments are addressed.

I've changed `Navigatable.getTargetUri` to `Navigatable.getResourceUri`. It should be implemented by widgets which handle some `Resource`, not necessary only `FileResource` as before. For example, it makes sense to reveal a corresponding file in the navigator and provide workspace variables based on a git resource as well. Probably the better name will be not `Navigatable` but `ResourceWidget`.

**Comment by svenefftinge** [comment]
> Probably the better name will be not Navigatable but ResourceWidget and ResourceWidgetOpenHandler. I would like to rename it, does it make sense to you?

Yes, maybe it's a bit better. Both names are fine with me.
